File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.py
import sys
input = lambda: sys.stdin.readline().strip()

def sol():
    n, k = map(int, input().split())
    bag = [[0, 0]]
    for _ in range(n):
        bag.append(list(map(int, input().split())))
    dp = [[0] * (k+1) for _ in range(n+1)]
    for i in range(1, n+1):
        for j in range(1, k+1):
            if j >= bag[i][0]:
                dp[i][j] = max(dp[i-1][j], dp[i-1][j-bag[i][0]] + bag[i][1])
            else:
                dp[i][j] = dp[i-1][j]
    print(dp[n][k])

sol()

File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {

	private static final BufferedReader  br = new BufferedReader(new InputStreamReader(System.in));
	private static       StringTokenizer st;

	private static int N;
	private static int K;

	private static int[]   knapsack;
	private static int[][] obj;

	public static void main(String[] args) throws IOException {
		init();
		sol();
	}

	private static void init() throws IOException {
		st = new StringTokenizer(br.readLine());

		N = Integer.parseInt(st.nextToken());
		K = Integer.parseInt(st.nextToken());

		obj = new int[N + 1][2];
		for (int i = 1; i <= N; i++) {
			st = new StringTokenizer(br.readLine());

			obj[i][0] = Integer.parseInt(st.nextToken());
			obj[i][1] = Integer.parseInt(st.nextToken());
		}

		knapsack = new int[K + 1];
	}

	private static void sol() {
		for (int i = 1; i <= N; i++) {
			for (int j = K; j >= obj[i][0]; j--) {
				knapsack[j] = Math.max(knapsack[j], knapsack[j - obj[i][0]] + obj[i][1]);
			}
		}
		System.out.println(knapsack[K]);
	}

}